tensorflow-op-codegen/src/main.rs

```diff
@@ -221,6 +227,181 @@ fn write_attr<W: Write>(w: &mut W, attr: &Attr) -> Result<(), io::Error> {
+    keywords: &HashSet<String>,
+) -> Result<(), io::Error> {
+    let mut escaper = Escaper::new(keywords);
+    let escaped_args: Vec<_> = args.iter().map(|arg| escaper.escape(&arg)).collect();
+    write!(w, "    fn build_impl_outputs(&self, ")?;
+    for arg in &escaped_args {
+        write!(w, "{}: crate::Output, ", arg)?;
+    }
+    let scope_var = escaper.escape("scope");
+    let node_var = escaper.escape("nd");
+    write!(
+        w,
+        r#"{scope}: &mut crate::Scope) -> crate::Result<{op_name}Outputs> {{
+        let op = {scope}.new_operation({op_name:?}, |{node}| {{
+"#,
+        scope = scope_var,
+        op_name = op_name,
+        node = node_var,
+    )?;
+    for arg in escaped_args {
+        write!(w, "            {}.add_input({});\n", node_var, arg)?;
+    }
+    write!(w, "            for op in &self.control_inputs {{\n")?;
+    write!(w, "                {}.add_control_input(op);\n", node_var)?;
+    write!(w, "            }}\n")?;
+    for attr in attrs {
+        write_set_attr(w, attr, &node_var)?;
+        //TODO: also get type and value of each number_attr by looking up the corresponding attribute
+    }
+    write!(w, "            ::std::result::Result::Ok(())\n")?;
+    write!(w, "        }})?;\n")?;
+    write!(w, "        Ok({}Outputs{{op}})\n", op_name)?;
+    write!(w, "    }}\n")?;
+    Ok(())
+}
+
+///writes the impl for the output struct that includes slicing implementations for Outputs that have output.number_attr set
+fn write_build_impl_outputs_struct<W: Write>(
+    w: &mut W,
+    op_name: &str,
+    attrs: &[Attr],
+    args: &[String],
+    outputs: Vec<Output>,
+    keywords: &HashSet<String>,
+) -> Result<(), io::Error> {
+    let mut escaper = Escaper::new(keywords);
+    let escaped_args: Vec<_> = args.iter().map(|arg| escaper.escape(&arg)).collect();
+
+    write!(w, "impl {}Outputs {{\n", op_name)?;
+
+    for (i, output) in outputs.iter().enumerate() {
+        if output.number_attr.is_some() {
+            let number_attr = output.number_attr.clone().unwrap();
+            write!(
+                w,
+                "    /// Returns an OutputSlice for the '{}' output of the {} operation.\n",
+                output.rust_name, op_name
+            )?;
+            write!(
+                w,
+                "    pub fn {}(&self) -> crate::graph::OutputSlice {{\n",
+                output.rust_name
+            )?;
+            //create an Output for this index
+            write!(w, "        let forward_output = crate::Output {{\n")?;
+            write!(w, "            operation: self.op.clone(),\n")?;
+            write!(w, "            index: {}\n", i)?;
+            write!(w, "        }};\n")?;
+            //now get that Outputs operation and create an OutputSlice given the runtime length
+            write!(
+                w,
+                r#"        crate::graph::OutputSlice::new(forward_output.operation.clone(), self.op.get_attr_int("{}").unwrap())"#,
```

Comment:
These are always starting at index 0, but that's not correct if this is not the first output arg.

Also, for both this and the `else` branch, the index needs to be adjusted for the size of prior outputs.  In other words, if the first `output_arg` has size 4, and the second has size 5, then the first needs to take outputs 0-3 and the second needs to take 4-8.

To do this, you'll probably need to read all the relevant int attributes as per the TODO above and either pass the `Output`s or the indices into the `FooOutputs` struct.

Reply:
thank you, patching this now.